infra/Pos/DHT/Real/Types.hs

```diff
@@ -10,28 +11,24 @@ import           Universum              hiding (fromStrict, toStrict)
 import           Control.Concurrent.STM (TVar)
 import qualified Data.ByteString        as BS
 
-import           Data.Store             (PeekException (..), decodeIOPortionWith)
 import qualified Network.Kademlia       as K
 
-import           Pos.Binary.Class       (Bi (..), encode)
+import           Data.Bifunctor         (bimap)
+import           Pos.Binary.Class       (Bi (..), serialize', decodeFull)
 import           Pos.DHT.Model.Types    (DHTData, DHTKey)
 import           Pos.Util.TimeWarp      (NetworkAddress)
-import           System.IO.Unsafe       (unsafePerformIO)
-
-fromBSBinary :: Bi b => BS.ByteString -> Either String (b, BS.ByteString)
-fromBSBinary bs = unsafePerformIO $
-    (decodeIOPortionWith get bs >>= \(off, res) -> return $ Right (res, BS.drop off bs))
-      `catch` handler
-  where
-    handler (PeekException {..}) = return $ Left (toString peekExMessage)
 
+-- CSL-1296: Should we worry about leftovers?
+-- Previous Store-based implementation was taking the leftover
+-- out of the serialisation, whereas in CBOR we deserialise everything
+-- in a gulp.
```

Comment:
This depends on how we're doing framing now with the network layer. If we're doing external framing with an overall length prefix and we collect that much data and then supply that all in one go, then it's essential that we check for and prohibit leftovers (as that would be a protocol error).

On the other hand, if we're now doing internal framing then we take the leftovers and we push that back into the channel, since it's part of the next message. So this requires checking, and you'll need help on that since it's not code you know.